pkg/tracing/config.go

```diff
@@ -0,0 +1,158 @@
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package tracing provides the canonical tracing functionality used by Go-based
+// Istio components.
+//
+// The package provides direct integration with the Cobra command-line processor which makes it
+// easy to build programs that use a consistent interface for tracing. Here's an example
+// of a simple Cobra-based program using this tracing package:
+//
+//		func main() {
+//			// get the default tracing options
+//			options := tracing.NewOptions()
+//
+//			rootCmd := &cobra.Command{
+//				Run: func(cmd *cobra.Command, args []string) {
+//
+//					// configure the tracing system
+//					if _, err := tracing.Configure(options); err != nil {
+//                      // print an error and quit
+//                  }
+//
+//					// Generate a trace
+//					ot.StartSpan(...)
+//				},
+//			}
+//
+//			// add tracing-specific flags to the cobra command
+//			options.AttachCobraFlags(rootCmd)
+//			rootCmd.SetArgs(os.Args[1:])
+//			rootCmd.Execute()
+//		}
+//
+// The point of this package is to configure the global OpenTracing tracer
+// for the process as read from ot.GlobalTracer and used by ot.StartSpan.
+package tracing
+
+import (
+	"fmt"
+	"io"
+	"time"
+
+	ot "github.com/opentracing/opentracing-go"
+	jaeger "github.com/uber/jaeger-client-go"
+	"github.com/uber/jaeger-client-go/transport"
+	"github.com/uber/jaeger-client-go/transport/zipkin"
+	"go.uber.org/zap"
+
+	"istio.io/istio/pkg/log"
+)
+
+type holder struct {
+	closer io.Closer
+	tracer ot.Tracer
+}
+
+var (
+	httpTimeout = 5 * time.Second
+	sampler     = jaeger.NewConstSampler(true)
+	poolSpans   = jaeger.TracerOptions.PoolSpans(false)
+	logger      = spanLogger{}
+)
+
+// indirection for testing
+type newZipkin func(url string, options ...zipkin.HTTPOption) (*zipkin.HTTPTransport, error)
+
+// Configure initializes Istio's tracing subsystem.
+//
+// You typically call this once at process startup.
+// Once this call returns, the tracing system is ready to accept data.
+func Configure(serviceName string, options *Options) (io.Closer, error) {
+	return configure(serviceName, options, zipkin.NewHTTPTransport)
+}
+
+func configure(serviceName string, options *Options, nz newZipkin) (io.Closer, error) {
+	reporters := make([]jaeger.Reporter, 0, 3)
+
+	if options.ZipkinURL != "" {
+		trans, err := nz(options.ZipkinURL, zipkin.HTTPLogger(logger), zipkin.HTTPTimeout(httpTimeout))
+		if err != nil {
+			return nil, fmt.Errorf("could not build zipkin reporter: %v", err)
+		}
+		reporters = append(reporters, jaeger.NewRemoteReporter(trans))
+	}
+
+	if options.JaegerURL != "" {
```

Comment:
note: i think that previous testing revealed a race in enabling both Zipkin and Jaeger at the same time. we might want to think about enforcing mutual exclusivity on the flags somehow.

Reply:
OK, I made the options mutually exclusive.